### PR TITLE
[FEATURE] New custom widget checkable combobox

### PR DESCRIPTION
**Pull request by alexbruy**
Add new custom widget: combobox that allow multiple selection. Hope it can be useful in some cases, for example in Processing where we now use modal dialogs for selecting options.

Some screenshots below.
checkable-combo checkable-resized-open
checkable-resized

There is one issue which should be solved: dropdown list closed after clicking on item, so to select multiple items one should open dropdown several times. Help in this direction is welcome.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by NathanW2** [comment]
My only main concern with this is from a usability point of view.   What
happens if I select dropdown and click on "Two" but not the check box?

**Comment by alexbruy** [comment]
@NathanW2 with current implementation item will selected or deselected (checked/unchecked).

File: src/customwidgets/qgscheckablecomboboxplugin.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/***************************************************************************
+   qgspasswordlineeditplugin.cpp
```
Comment (nyalldawson) [inline src/customwidgets/qgscheckablecomboboxplugin.cpp]: Needs updating

File: src/gui/qgscheckablecombobox.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  QStyleOptionViewItem &nonConstOpt = const_cast<QStyleOptionViewItem &>( option );
+  nonConstOpt.showDecorationSelected = false;
+  QStyledItemDelegate::paint( painter, nonConstOpt, index );
+}
+
+
+QgsCheckableComboBox::QgsCheckableComboBox( QWidget *parent )
+  : QComboBox( parent )
+  , mSeparator( ", " )
+  , mDefaultText( "" )
```
Comment (nyalldawson) [inline src/gui/qgscheckablecombobox.cpp]: Not required

File: src/gui/qgscheckablecombobox.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+{
+  setModel( new QgsCheckableItemModel( this ) );
+  setItemDelegate( new QgsCheckBoxDelegate( this ) );
+
+  QLineEdit *lineEdit = new QLineEdit( this );
```
Comment (nyalldawson) [inline src/gui/qgscheckablecombobox.cpp]: I'm wondering... (Might be overkill) what about a filter line edit here so we get a way to instantly clear the checked values?

File: src/gui/qgscheckablecombobox.h [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+#include "qgis_gui.h"
+
+/** \class QgsCheckableItemModel
+ * \ingroup gui
+ * QStandardItemModel subclass which makes all items checkable
+ * by default.
+ * @note added in QGIS 3.0
+ * @note not available in Python bindings
+ **/
+class GUI_EXPORT QgsCheckableItemModel : public QStandardItemModel
```
Comment (nyalldawson) [inline src/gui/qgscheckablecombobox.h]: I'd just mark this as private API so it's not subject to to freeze

File: src/gui/qgscheckablecombobox.h [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+/** \class QgsCheckBoxDelegate
+ * \ingroup gui
+ * QStyledItemDelegate subclass for QgsCheckableComboBox. Needed for
+ * correct drawing of the checkable items on Mac and GTK.
+ * @note added in QGIS 3.0
+ * @note not available in Python bindings
+ **/
+
+class GUI_EXPORT QgsCheckBoxDelegate : public QStyledItemDelegate
```
Comment (nyalldawson) [inline src/gui/qgscheckablecombobox.h]: Same here - I'd go private API

**Comment by nyalldawson** [comment]
I like the idea! We've already got this used in some places (pg source select for pk, I think also in dbmanager query window), so it'll be nice to remove the duplicate code and have one well tested implementation.

**Comment by nirvn** [comment]
This could be useful in the save symbol dialog, where we currently use a line edit widget for tags.

File: src/gui/qgscheckablecombobox.cpp (hunk 3 above)
Comment (alexbruy) [inline src/gui/qgscheckablecombobox.cpp]: I was thinking about adding context menu to the dropdown with two options: "select all" and "clear selection"

**Comment by NathanW2** [comment]
@alexbruy good to merge?

**Comment by alexbruy** [comment]
Added context menu with select all/deselect all options, fixed popup behaviour. Also make SIP bindings autogenerated.

**Comment by alexbruy** [comment]
@NathanW2 should be fine now.

File: src/gui/qgscheckablecombobox.h [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    Q_PROPERTY( QString separator READ separator WRITE setSeparator )
+    Q_PROPERTY( QString defaultText READ defaultText WRITE setDefaultText )
+    Q_PROPERTY( QStringList checkedItems READ checkedItems WRITE setCheckedItems )
+
+  public:
+
+    /** Constructor for QgsCheckableComboBox.
+     * \param parent parent object
+     */
+    QgsCheckableComboBox( QWidget *parent = nullptr );
```
Comment (nyalldawson) [inline src/gui/qgscheckablecombobox.h]: Technically speaking this should be 

     QgsCheckableComboBox( QWidget *parent SIP_TRANSFERTHIS = nullptr );

But I don't think it makes any real world difference

File: tests/src/python/test_qgscheckablecombobox.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+__copyright__ = 'Copyright 2017, The QGIS Project'
+# This will get replaced with a git SHA1 when you do a git archive
+__revision__ = '$Format:%H$'
+
+import qgis  # NOQA
+
+from qgis.PyQt.QtCore import Qt
+from qgis.gui import QgsCheckableComboBox
+
+try:
```
Comment (nyalldawson) [inline tests/src/python/test_qgscheckablecombobox.py]: This shouldn't be required anymore - it was only needed when we still supported Qt4 and QSignalSpy wasn't available. It's always present for Qt5 builds

**Review by nyalldawson — APPROVED** [review]
Looks great to me... very thorough and careful code with fantastic attention to detail

**Pull request merged**